File: AE/Arrays/smallest_difference.py
"""
- input:
  - two non-empty arrays of integers
- output:
  - return the pair of numbers (one from each array) whose absolute difference is closest to zero 
  - return the pair in an array, with the number from the first array in the first position
- logic:
  - the goal is to efficiently look at the pairs to determine which has the smallest difference
  - we know that the difference between two numbers gets smaller,  as the smaller number gets closer to the larger number
  - therefore, we can sort the two arrays --> numbers are in increasing order so the next number in the array will be greater than the current number
  - have a pointer for both arrays
  - use the pointers to calculate the absolute difference and track the smallest 
  - and then since we have sorted the numbers in ascending order, determine which of the two numbers is smaller and increment its pointer by 1
  - in doing so, we are essentially moving our position to a larger number and that larger number will reduce the difference between the two numbers
"""


def smallest_difference(arrayOne, arrayTwo):
    arrayOne.sort()
    arrayTwo.sort()
    pair = []
    smallest_diff = float('inf')
    pointer_one = 0
    pointer_two = 0

    while pointer_one < len(arrayOne) and pointer_two < len(arrayTwo):
        num1 = arrayOne[pointer_one]
        num2 = arrayTwo[pointer_two]

        absolute_diff = abs(num1 - num2)

        if absolute_diff < smallest_diff:
            smallest_diff = absolute_diff
            pair = [num1, num2]

        if num1 < num2:
            pointer_one += 1
        else:
            pointer_two += 1

    return pair

# Time Complexity: O(nlogn + mlogm)
# Space Complexity: O(1)

File: AE/Arrays/smallestDifference.js
/*
Write a function that takes in two non-empty arrays of integers, finds the pair of numbers (one from each array) whose absolute difference is closest to zero, and returns an array containing these two numbers, with the number from the first array in the first position.

Note that the absolute difference of two integers is the distance between them on the real number line. For example, the absolute difference of -5 and 5 is 10 , and the absolute difference of -5 and -4 is 1

Sample Input:
arrayOne = [-1,5,10,20,28,3]
arrayTwo = [26,134,135,15,17]

Sample Output:
[28,26]
*/


const smallestDifferenceNaive = (arrayOne, arrayTwo) => {
  let smallestDiff = Infinity;
  let result = [];

  for (let i = 0; i < arrayOne.length; i++) {
    const num1 = arrayOne[i];
    for (let j = 0; j < arrayTwo.length; j++) {
      const num2 = arrayTwo[j];

      const absDiff = Math.abs(num1 - num2);

      if (absDiff < smallestDiff) {
        smallestDiff = absDiff;
        result = [num1, num2];
      }
    }
  }
  return result;
}

// Time Complexity: O(N^2);
// Space Complexity: O(1);

const smallestDifferenceOptimized = (arrayOne, arrayTwo) => {
  let smallestDiff = Infinity;
  let result = [];

  // by sorting the search for the smallest difference begins with the smallest elements from both the respective inputs
  // and unnecessary operations can be avoided
  arrayOne.sort((a,b) => a - b);
  arrayTwo.sort((a,b) => a - b);

  let idx1 = 0;
  let idx2 = 0;

  while (idx1 < arrayOne.length && idx2 < arrayTwo.length) {
    const num1 = arrayOne[idx1];
    const num2 = arrayTwo[idx2];

    const absDiff = Math.abs(num1 - num2);
    if (absDiff < smallestDiff) {
      smallestDiff = absDiff;
      result = [num1, num2];
    }

    // searching for the smalelst difference, so close the gap by moving the pointer of the smaller number
    if (num1 < num2) {
      idx1 += 1
    } else {
      idx2 += 1;
    }
  }
  return result;
}

// Time Complexitty: O(NlogN + MlogM);
// Space Complexity: O(1);